packages/workbox-routing/lib/Router.mjs

```diff
@@ -0,0 +1,328 @@
+/*
+ Copyright 2017 Google Inc. All Rights Reserved.
+ Licensed under the Apache License, Version 2.0 (the "License");
+ you may not use this file except in compliance with the License.
+ You may obtain a copy of the License at
+
+     http://www.apache.org/licenses/LICENSE-2.0
+
+ Unless required by applicable law or agreed to in writing, software
+ distributed under the License is distributed on an "AS IS" BASIS,
+ WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ See the License for the specific language governing permissions and
+ limitations under the License.
+*/
+
+import core from 'workbox-core';
+import {_private} from 'workbox-core';
+
+import normalizeHandler from './normalizeHandler.mjs';
+
+/**
+ * The Router takes one or more [Routes]{@link Route} and allows you to apply
+ * that routing logic to determine the appropriate way of handling requests
+ * inside of a service worker.
+ *
+ * It also allows you to define a "default" handler that applies to any requests
+ * that don't explicitly match a `Route`, and a "catch" handler that responds
+ * to any requests that throw an exception while being routed.
+ *
+ * You can use the `handleRequest()` method to pass a `FetchEvent` through the
+ * router and ultimately get a "routed response" back.
+ * If you'd like this to be handled automatically for you, calling
+ * `addFetchListener()` will cause the `Router` to respond to `fetch` events.
+ *
+ * If a request matches multiple routes, precedence will be given to the
+ * **earliest** registered route.
+ *
+ * @memberof module:workbox-routing
+ */
+class Router {
+  /**
+   * Constructs a new `Router` instance, without any registered routes.
+   */
+  constructor() {
+    // _routes will contain a mapping of HTTP method name ('GET', etc.) to an
+    // array of all the corresponding Route instances that are registered.
+    this._routes = new Map();
+    this._isListenerRegistered = false;
+  }
+
+  /**
+   * This will register a `fetch` event listener on your behalf which will check
+   * the incoming request to see if there's a matching route, and only respond
+   * if there is a match.
+   *
+   * @return {boolean} Returns `true` if this is the first time the method is
+   * called and the listener was registered. Returns `false` if called again,
+   * as the listener will only be registered once.
+   */
+  addFetchListener() {
+    if (!this._isListenerRegistered) {
+      this._isListenerRegistered = true;
+      self.addEventListener('fetch', (event) => {
+        const responsePromise = this.handleRequest(event);
```

Comment:
Is handleRequest synchronous? i.e. we know if a Route is going to handle this and so we don't need to call event.waitUntil?

Reply:
Yup, `handleRequest` is synchronous.

It will return a `Promise<Response>` if there's a matching route, or a default route, but it doesn't do anything like `await` or call `.then()` on the promise.